src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,189 @@
+              arg.resize(1,
+                         GroupConcatAggregationData{
+                             aggregateInfo.aggregateType_.separator_.value()});
+            } else {
+              arg.resize(1);
+            }
+          },
+          aggregationData_.getAggregationDataVariant(
+              aggregateInfo.aggregateDataIndex_));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::resetAggregationData() {
+  for (const auto& alias : aggregateAliases_) {
+    for (const auto& aggregateInfo : alias.aggregateInfo_) {
+      std::visit([]<SupportedAggregates T>(T& arg) { arg.at(0).reset(); },
+                 aggregationData_.getAggregationDataVariant(
+                     aggregateInfo.aggregateDataIndex_));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::commitRow(
+    IdTable& resultTable,
+    sparqlExpression::EvaluationContext& evaluationContext,
+    const GroupBy& groupBy) {
+  resultTable.emplace_back();
+  for (size_t rowIndex = 0; rowIndex < groupByCols_.size(); ++rowIndex) {
+    resultTable.back()[rowIndex] = currentGroupBlock_.at(rowIndex).second;
+  }
+
+  evaluationContext._beginIndex = resultTable.size() - 1;
+  evaluationContext._endIndex = resultTable.size();
+
+  for (auto& alias : aggregateAliases_) {
+    groupBy.evaluateAlias(alias, &resultTable, evaluationContext,
+                          aggregationData_, &localVocab_);
+  }
+  resetAggregationData();
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::processNextBlock(
+    sparqlExpression::EvaluationContext& evaluationContext, size_t beginIndex,
+    size_t endIndex) {
+  size_t blockSize = endIndex - beginIndex;
+  evaluationContext._beginIndex = beginIndex;
+  evaluationContext._endIndex = endIndex;
+
+  for (auto& aggregateAlias : aggregateAliases_) {
+    for (auto& aggregate : aggregateAlias.aggregateInfo_) {
+      // Evaluate child expression on block
+      auto exprChildren = aggregate.expr_->children();
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
+
+      std::visit(
+          [this, blockSize, &evaluationContext,
+           &aggregate]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), blockSize, &evaluationContext);
+
+            for (const auto& val : generator) {
+              addToAggregateFunction(aggregate.aggregateDataIndex_, val,
+                                     &evaluationContext);
```

Comment:
If this function call contains a `std::visit` over the different aggregate types, then this loop is inverted, 
The `visit`on the aggregate should be outside, and the visit on the `ExpressionResult` should be on the inside
(more efficient for large blocks).